src/Microsoft.Extensions.OData.Migration/Formatters/Deserialization/ODataMigrationInputFormatter.cs

```diff
@@ -0,0 +1,257 @@
+                return Task.FromResult(InputFormatterResult.Failure());
+            }
+        }
+
+        /// <summary>
+        /// Read and deserialize incoming object from HTTP request stream.
+        /// </summary>
+        /// <param name="type">incoming request body object type.</param>
+        /// <param name="defaultValue">default value for this type.</param>
+        /// <param name="model">Edm model to reference when validating.</param>
+        /// <param name="baseAddress">Base address of request.</param>
+        /// <param name="internalRequest">HTTP request method that contains ODataPath.</param>
+        /// <param name="getODataRequestMessage">Function to obtain customized ODataMessageWrapper.</param>
+        /// <param name="getEdmTypeDeserializer">Function to obtain appropriate edm deserializer.</param>
+        /// <param name="getODataPayloadDeserializer">Function to obtain appropriate deserialize for function payloads.</param>
+        /// <param name="getODataDeserializerContext">Context for Deserializer.</param>
+        /// <param name="registerForDisposeAction">Registration function for disposables.</param>
+        /// <returns>Deserialized object.</returns>
+        internal object ReadFromStream(
+            Type type,
+            object defaultValue,
+            IEdmModel model,
+            Uri baseAddress,
+            HttpRequest internalRequest,
+            Func<IODataRequestMessage> getODataRequestMessage,
+            Func<IEdmTypeReference, ODataDeserializer> getEdmTypeDeserializer,
+            Func<Type, ODataDeserializer> getODataPayloadDeserializer,
+            Func<ODataDeserializerContext> getODataDeserializerContext,
+            Action<IDisposable> registerForDisposeAction,
+            Action<Exception> logErrorAction)
+        {
+            object result;
+
+            IEdmTypeReference expectedPayloadType;
+            ODataPath path = internalRequest.ODataFeature().Path;
+            ODataDeserializer deserializer = GetDeserializer(type, path, model, getEdmTypeDeserializer, getODataPayloadDeserializer, out expectedPayloadType);
+            if (deserializer == null)
+            {
+                throw new ArgumentException("type", "Formatter does not support reading type " + type.FullName);
+            }
+
+            try
+            {
+                ODataMessageReaderSettings oDataReaderSettings = internalRequest.GetReaderSettings();
+                oDataReaderSettings.BaseUri = baseAddress;
+                oDataReaderSettings.Validations = oDataReaderSettings.Validations & ~ValidationKinds.ThrowOnUndeclaredPropertyForNonOpenType;
+
+                IODataRequestMessage oDataRequestMessage = getODataRequestMessage();
+                ODataMessageReader oDataMessageReader = new ODataMessageReader(oDataRequestMessage, oDataReaderSettings, model);
+                registerForDisposeAction(oDataMessageReader);
```

Comment:
I don't understand this.  Does the oDataMessageReader really need to live longer than the scope of this method (i.e. it can't be disposed right after the call to deserializer.Read)?